### ISSUE TITLE
Incorrect .throw() result on generator with yield* to another generator with try/catch

### ISSUE THREAD

**Issue by bterlson** [Microsoft/ChakraCore#137]
Repro:

```js
function* test() {
  try {
    yield;
  } catch (e) {
    return;
  }
}

function* wrap() {
  yield* test();
  yield 3;
}

var iter = wrap();
iter.next();
print(JSON.stringify(iter.throw(Error("fail"))));
```

Expected per spec is `iterator result object { value: undefined, done: true }` as yield* has a return completion value when received is a throw completion. However, v8/sm do differently and https://github.com/tc39/ecma262/issues/293 proposes changing the expected result to `{value: 3, done: false }` by returning a normal completion for this case.

**Comment by bterlson** [comment]
The v8/sm semantics won the day here, so the expected result is `{value: 3, done: false}`.

**Comment by abchatra** [comment]
This is failing a kangax test case as well. 
```js
print((function()
{
function * generator() {
  yield * (function * () {
    try {
      yield 'foo';
    }
    catch(e) {
      return;
    }
  }());
  yield 'bar';
}
var iter = generator();
iter.next();
return iter['throw']().value === 'bar';
})());
```

**Comment by nicolo-ribaudo** [comment]
Can I tackle this issue?

**Comment by ianwjhalliday** [comment]
@nicolo-ribaudo the 1.2 milestone bugs need to be fixed asap.  I'm happy to let you take a stab at it.  Please start on it immediately and let me know within the next couple of days what your plan for the fix is.

Let me know if you do not have enough time to complete this bug fix within the next week or so.  I will want to start on it first thing next week if that is the case.

**Comment by ianwjhalliday** [comment]
@nicolo-ribaudo also note that the due date on the 1.2 milestone is a rough estimate.  Lockdown on our bug fixing for the next release may occur sooner which is why this needs to be addressed asap.

**Comment by nicolo-ribaudo** [comment]
@ianwjhalliday I'm working on this now. Is it ok to move the call to the iterator's `.next()` method from the bytecode to the `JavascriptOperators::OP_ResumeYield` function? It probably will be a little less efficient, but I don't know if there is a way to know in the bytecode how the generator has been resumed.

**Comment by nicolo-ribaudo** [comment]
Should this code throw?
```js
function* gen() {
  yield* {
    [Symbol.iterator]() {
      return {
        next() { return { value: 3, done: false } },
        throw() { return 11; } // <-- 11 is not an object!!!
      }
    }
  };
};
var g = gen();
g.next();
g.throw(3);
```
Both Chrome and Firefox don't throw, and [there is a test which ensures that this code doesn't throw](https://github.com/Microsoft/ChakraCore/blob/master/test/es6/generators-functionality.js#L1374-L1376), but the spec says:

```
YieldExpression: yield*AssignmentExpression
...
5.b. Else if received.[[Type]] is throw, then
    i. Let throw be ? GetMethod(iterator, "throw").
   ii. If throw is not undefined, then
        1. Let innerResult be ? Call(throw, iterator, « received.[[Value]] »).
        2. NOTE: Exceptions from the inner iterator throw method are propagated. Normal completions from an inner throw method are processed similarly to an inner next.
        3. If Type(innerResult) is not Object, throw a TypeError exception.
```

**Comment by ianwjhalliday** [comment]
Yes, looks like that should throw `TypeError`.  Chrome and Firefox probably have the same bug.

Whoa.  I just read [14.4.14](https://tc39.github.io/ecma262/#sec-generator-function-definitions-runtime-semantics-evaluation) for `yield*` evaluation.  The spec has changed a lot since @aneeshdk and I implemented `yield*`.  E.g. if the inner iterator does not have a `throw` method then we are supposed to throw a `TypeError` due to `yield*` protocol violation.  This is completely unfamiliar behavior to me, but it has been a long while now since I've read this part of the spec.

Sounds like our implementation of `yield*` needs to be redesigned to match the spec.

@nicolo-ribaudo sounds like you're investigating test cases to see the various browsers` behaviors versus the spec.  That's good.  What I would do in this case if I were tackling this right now is review our current test cases and update and add to them to reflect the new spec.  From there I would identify the different patterns that we are failing to implement, and anything we have implemented that is incorrect now.  With that knowledge I would finally visit designing the new implementation.

How do you feel about taking that approach?  Might as well also vet the `return()` behavior at the same time.

**Comment by nicolo-ribaudo** [comment]
PR: #597

**Comment by ianwjhalliday** [comment]
Fixed by #665
